R/CrosswalksFromATTAINS.R

```diff
@@ -0,0 +1,425 @@
+#' OrganizationIdentifier, OrganizationFormalName, ProviderName.
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' provider.ref <- TADA_GetProviderRef()
+#' }
+#'
+TADA_GetProviderRef <- function() {
+  # If there is a cached table available return it
+  if (!is.null(WQXProviderRef_Cached)) {
+    return(WQXProviderRef_Cached)
+  }
+
+  # Try to download up-to-date raw data
+  raw.data <- tryCatch(
+    {
+      # read raw csv from url
+      utils::read.csv(url("https://www.waterqualitydata.us/data/Organization/search?mimeType=csv&zip=no")) %>%
+        dplyr::select(OrganizationIdentifier, OrganizationFormalName, ProviderName) %>%
+        dplyr::distinct()
+    },
+    error = function(err) {
+      NULL
+    }
+  )
+
+  # If the download failed fall back to internal data (and report it)
+  if (is.null(raw.data)) {
+    message("Downloading latest WQP Organization and Provider Reference Table failed!")
+    message("Falling back to (possibly outdated) internal file.")
+    return(utils::read.csv(system.file("extdata", "WQXProviderRef.csv", package = "EPATADA")))
+  }
+
+  # Save updated table in cache
+  WQXProviderRef <- raw.data
+
+  WQXProviderRef_Cached <- WQXProviderRef
+
+  WQXProviderRef
+}
+
+# Update Characteristic Reference Table internal file (for internal use only)
+
+TADA_UpdateProviderRef <- function() {
+  utils::write.csv(TADA_GetProviderRef(), file = "inst/extdata/WQXProviderRef.csv", row.names = FALSE)
+}
+
+
+#' Update Monitoring Location Identifiers in ATTAINS
+#'
+#' This function creates the batch upload files needed to update the 
+#' Monitoring Location Identifiers in ATTAINS Assessment Unit profiles. Users
+#' can specify whether all records should be overwritten or if new 
+#' Monitoring Location Identifiers should be appended to existing records.
+#'
+#' @param org_id Character argument. The ATTAINS organization identifier must  
+#' be supplied by the user. A list of organization identifiers can be found by
+#' downloading the ATTAINS Domains Excel file: 
+#' https://www.epa.gov/system/files/other-files/2023-09/DOMAINS.xlsx. 
+#' Organization identifiers are listed in the "OrgName" tab. The "code" column
+#' contains the organization identifiers that should be used for this param.
+#'
+#' @param data_links Character argument. When data_links is equal to "update" 
+#' or "replace", the function will build the URL for the Water Quality Portal 
+#' Data Site page for each Monitoring Location in the df. It will examine the 
+#' response codes of the URLs and only retain those with a 200 response, which 
+#' indicates they are valid urls. When data_links = "update", the url will be 
+#' added to any existing text in the MS_DATA_LINK_TEXT column. When 
+#' data_links = "replace", the url will replace any existing text in the
+#' MS_DATA_LINK_TEXT column. When data_links = "none", no URLs will be created
+#' or added to the df. Default is data_links = "update".
```

Comment:
I changed "update" to "add" for this function input because update was confusing to me and felt too similar of a term to the other "replace" option.

Reply:
I also change data_links to wqp_data_links for clarity